### PR TITLE
EOS-26021: S3 Audit Log: Expose an API to push audit logs

### PR DESCRIPTION
**Pull request by vastradparayya**
# Problem Statement
- S3 Audit Log: Expose an API to push audit logs

# Design
-  https://jts.seagate.com/browse/EOS-26021

# Coding 
-  Coding conventions are followed and code is consistent [Y/N]: Y
-  Confirm All CODACY errors are resolved [Y/N]: Y

# Testing 
- [x] Confirm that Test Cases are added (for both the cases, fix and feature) [Y/N]: Y
- [x] Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability [Y/N]: Y
- [x] Confirm Testing was performed with installed RPM [Y/N]:  Y

# Review Checklist 
  Before posting the PR please ensure
- [x] PR is self reviewed
- [x] Is there a change in filename/package/module or signature [Y/N]: N
- [x] If yes for above point, Is a notification sent to all other cortx components [Y/N] NA
- [x] Jira is updated
- [x] Check if the description is clear and explained. 
- [x] Check Acceptance Criterion is defined. 
- [x] All the tests performed should be mentioned before Resolving a JIRA. 
- [x] Verification needs to be done before marked as Closed/Verified 

# Documentation
- [ ] Changes done to WIKI / Confluence page

### CODE REVIEW

File: py-utils/src/utils/audit_log/error.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+
+class AuditLogError(Exception):
+    """Generic Exception with error code and output."""
```
Comment (rohit-k-dwivedi) [inline py-utils/src/utils/audit_log/error.py:19]: inherit UtilsError to avoid redefining class

Comment (vastradparayya) [reply py-utils/src/utils/audit_log/error.py]: Done

File: py-utils/src/utils/audit_log/audit_log_server.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+class AuditLogRequestHandler(RestServer):
+    """Rest interface of Audit log."""
+    @staticmethod
+    async def receive(request):
+        Log.debug("Received POST request for audit message")
+        try:
+            payload = await request.json()
+            messages = payload['messages']
+            # TODO Store audit logs messages to a persistent storage.
+        except AuditLogError as e:
```
Comment (rohit-k-dwivedi) [inline py-utils/src/utils/audit_log/audit_log_server.py:38]: why would the above try block raise AuditlogError

Comment (vastradparayya) [reply py-utils/src/utils/audit_log/audit_log_server.py:38]: Yes @rohit-k-dwivedi, there is storing messages into persistent storage(TODO)

File: py-utils/src/utils/audit_log/error.py (hunk 1 above)
Comment (rahul-tripathi-git) [reply py-utils/src/utils/audit_log/error.py]: Marking this comment as resolved,

File: py-utils/src/utils/utils_server/utils_server.py [hunk 3]
```diff
@@ -29,12 +29,16 @@ def __init__(self):
...
+        from cortx.utils.audit_log import AuditLogRequestHandler
         app.add_routes([web.post('/EventMessage/event', IemRequestHandler.send), \
             web.get('/EventMessage/event', IemRequestHandler.receive), \
             web.post('/MessageBus/message/{message_type}', \
             MessageBusRequestHandler.send), \
             web.get('/MessageBus/message/{message_type}', \
-            MessageBusRequestHandler.receive)])
+            MessageBusRequestHandler.receive),
+            web.post('/AuditLog/message/', \
+            AuditLogRequestHandler.receive)
```
Comment (sachinpunadikar) [inline py-utils/src/utils/utils_server/utils_server.py:40]: Should it be AuditLogRequestHandler.send ? For every post request, we have associated method ending with send

File: py-utils/test/audit_log/test_audit_log_server.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import json
+import unittest
+import requests
+
+
+class TestMessage(unittest.TestCase):
+    """Test AuditLog rest server functionality."""
+    _base_url = 'http://127.0.0.1:28300/AuditLog/message/'
+
+    def test_post(self):
```
Comment (selvakumaar5496) [inline py-utils/test/audit_log/test_audit_log_server.py:29]: tchange it to test_send

**Review by sachinpunadikar — APPROVED** [review]
+1

**Review by rohit-k-dwivedi — APPROVED** [review]
LGTM

File: py-utils/test/audit_log/test_audit_log_server.py (hunk 4 above)
Comment (rahul-tripathi-git) [reply py-utils/test/audit_log/test_audit_log_server.py]: changed it to test_send_audit_log as discussed with @sachinpunadikar.

**Review by rahul-tripathi-git — APPROVED** [review]
Looks good to me.

**Review by selvakumaar5496 — APPROVED** [review]
LGTM

**Pull request merged**